app/models/reservation.rb

```diff
@@ -1,7 +1,20 @@
 class Reservation < ApplicationRecord
   belongs_to :seance
-  belongs_to :user
+  belongs_to :user, optional: true
+
+  before_validation :set_user
 
   validates :status, presence: true
+  validates :email, presence: true
   enum status: { reserved: 0, confirmed: 1, canceled: 2 }
+
+  private
+
+  def set_user
+    self.user_id = find_user
+  end
+
+  def find_user
+    User.where(email:).pluck(:id).first
+  end
```

Comment:
The method could be confusing, the **name** suggests that it will return an object of `User`, but returns the` id` of the user. 

```suggestion
  def find_user_id_by_email
    User.where(email:).pluck(:id).first
  end
```

BTW `User.where(email:).pluck(:id).first` it's a syntactic sugar for: `User.where(email: email).pluck(:id).first` which will be easier to read by newcomers IMHO.

Another thing is a question: Do you really think that process of finding the matched user to the reservation's email should be placed in this model? I don't want to judge it, potential pros&cons of this decision you will see soon I suppose  ;)